src/Streamly/Internal/Data/Fold/Concurrent/Channel.hs

```diff
@@ -0,0 +1,91 @@
+-- |
+-- Module      : Streamly.Internal.Data.Fold.Concurrent.Channel
+-- Copyright   : (c) 2022 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+
+module Streamly.Internal.Data.Fold.Concurrent.Channel
+    (
+    -- * Channel
+      Channel
+
+    -- * Configuration
+    , Config
+    , maxBuffer
+    , inspectMode
+
+    -- * Fold operations
+    , evalWith
+    )
+where
+
+import Control.Concurrent (takeMVar)
+import Control.Monad (void, when)
+import Control.Monad.IO.Class (MonadIO(liftIO))
+import Data.IORef (writeIORef)
+import Streamly.Internal.Control.Concurrent (MonadAsync)
+import Streamly.Internal.Data.Fold (Fold(..), Step (..))
+import Streamly.Internal.Data.Stream.Channel.Worker (sendWithDoorBell)
+import Streamly.Internal.Data.Time.Clock (Clock(Monotonic), getTime)
+
+import Streamly.Internal.Data.Fold.Concurrent.Channel.Type
+import Streamly.Internal.Data.Stream.Channel.Types
+
+-------------------------------------------------------------------------------
+-- Evaluating a Fold
+-------------------------------------------------------------------------------
+
+-- XXX Cleanup the fold if the stream is interrupted. Add a GC hook.
+
+-- | Evaluate the fold asynchronously in a worker thread separate from the
+-- driver thread.
+--
+{-# INLINABLE evalWith #-}
+evalWith :: MonadAsync m => (Config -> Config) -> Fold m a b -> Fold m a b
+evalWith modifier f =
+    Fold step initial extract
+
+    where
+
+    initial = Partial <$> newChannel modifier f
+
+    -- XXX This is not truly asynchronous. If the fold is done we only get to
+    -- know when we send the next input unless the stream ends. We could
+    -- potentially throw an async exception to the driver to inform it
+    -- asynchronously. Alternatively, the stream should not block forever, it
```

Comment:
Throwing an async exception looks like a `callback`. Is it possible to design a `callback` like model? Basically `JUMP`. The exception does the job, I guess. But can we do it without an exception?

Polled stream abstraction looks clean though.